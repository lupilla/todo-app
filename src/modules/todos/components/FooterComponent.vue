<template>
  <footer class="footer" v-if="getAllTodos">
    <div class="todo-count">
      <strong v-if="getActiveTodos === 1">
        {{ getActiveTodos }} item left
      </strong>
      <strong v-else> {{ getActiveTodos }} items left </strong>
    </div>
    <ul class="filters">
      <li>
        <router-link
          :data-id-test="ALL"
          to="/"
          @click="selectedEvent(ALL)"
          :class="{ selected: linkSelected === ALL }"
        >
          All
        </router-link>
        <router-link
          :data-id-test="ACTIVE"
          :to="`/${ACTIVE}`"
          @click="selectedEvent(ACTIVE)"
          :class="{ selected: linkSelected === ACTIVE }"
        >
          Active
        </router-link>
        <router-link
          :data-id-test="COMPLETED"
          :to="`/${COMPLETED}`"
          @click="selectedEvent(COMPLETED)"
          :class="{ selected: linkSelected === COMPLETED }"
        >
          Completed
        </router-link>
      </li>
    </ul>
    <button
      type="button"
      data-id-test="remove"
      class="clear-completed"
      @click="removeTodos"
      v-if="getCompletedTodos"
    >
      Clear completed
    </button>
  </footer>
</template>
<script>
import { ALL, COMPLETED, ACTIVE } from "@/modules/todos/helpers/constants.js";
import { mapGetters, mapActions } from "vuex";
export default {
  name: "FooterComponent",
  data() {
    return {
      linkSelected: ALL,
    };
  },
  props: {
    todos: {
      type: Array,
      default: () => [],
    },
  },
  computed: {
    ...mapGetters("todos", ["allTodos", "activeTodos", "completedTodos"]),
    getActiveTodos() {
      return this.activeTodos?.length;
    },
    getCompletedTodos() {
      return this.completedTodos?.length;
    },
    getAllTodos() {
      return this.allTodos?.length;
    },
  },
  methods: {
    ...mapActions("todos", ["deleteCompletedTodos"]),
    async removeTodos() {
      try {
        await this.deleteCompletedTodos();
      } catch (error) {
        console.error(error);
      }
    },
    selectedEvent(todoLink) {
      this.linkSelected = todoLink;
    },
  },
  created() {
    this.ALL = ALL;
    this.ACTIVE = ACTIVE;
    this.COMPLETED = COMPLETED;
  },
  mounted() {
    if (this.$route.path === "/") {
      this.linkSelected = ALL;
    } else {
      this.linkSelected = this.$route.path.substring(1);
    }
  },
};
</script>
<style>
.footer {
  color: #777;
  padding: 10px 15px;
  height: 20px;
  text-align: center;
  border-top: 1px solid #e6e6e6;
}

.footer:before {
  content: "";
  position: absolute;
  right: 0;
  bottom: 0;
  left: 0;
  height: 50px;
  overflow: hidden;
  box-shadow: 0 1px 1px rgba(0, 0, 0, 0.2), 0 8px 0 -3px #f6f6f6,
    0 9px 1px -3px rgba(0, 0, 0, 0.2), 0 16px 0 -6px #f6f6f6,
    0 17px 2px -6px rgba(0, 0, 0, 0.2);
}

.todo-count {
  float: left;
  text-align: left;
}

.todo-count strong {
  font-weight: 300;
}

.filters {
  margin: 0;
  padding: 0;
  list-style: none;
  position: absolute;
  right: 0;
  left: 0;
}

.filters li {
  display: inline;
}

.filters li a {
  color: inherit;
  margin: 3px;
  padding: 3px 7px;
  text-decoration: none;
  border: 1px solid transparent;
  border-radius: 3px;
}

.filters li a:hover {
  border-color: rgba(175, 47, 47, 0.1);
}

.filters li a.selected {
  border-color: rgba(175, 47, 47, 0.2);
}

.clear-completed,
.clear-completed:active {
  float: right;
  position: relative;
  line-height: 20px;
  text-decoration: none;
  cursor: pointer;
}

.clear-completed:hover {
  text-decoration: underline;
}

@media (max-width: 430px) {
  .footer {
    height: 50px;
  }

  .filters {
    bottom: 10px;
  }
}
</style>
