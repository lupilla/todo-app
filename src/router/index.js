import { createRouter, createWebHistory } from "vue-router";

/**
 * Array of routes included in the application
 * - Default route ("/"): all todos are showed
 * - Active route ("/active"): only active todos are showed
 * - Completed route ("/completed"): only completed todos are showed
 * - Page not found route: error page 404 is showed when the route does not match
 */
const routes = [
  {
    path: "/",
    name: "allTodos",
    component: () =>
      import(
        /* webpackChunkName: 'alltodos'*/ "@/modules/todos/pages/TodoPage.vue"
      ),
  },
  {
    path: "/active",
    name: "activeTodos",
    component: () =>
      import(
        /* webpackChunkName: 'active'*/ "@/modules/todos/pages/TodoPage.vue"
      ),
  },
  {
    path: "/completed",
    name: "completedTodos",
    component: () =>
      import(
        /* webpackChunkName: 'completed'*/ "@/modules/todos/pages/TodoPage.vue"
      ),
  },
  {
    path: "/:pathMatch(.*)*",
    component: () =>
      import(
        /* webpackChunkName: 'NoPageFoundPage'*/ "@/modules/shared/pages/NoPageFound.vue"
      ),
  },
];

/**
 * Router creation method to export it to the main.js file
 */
const router = createRouter({
  history: createWebHistory(process.env.BASE_URL),
  routes,
});

export default router;
