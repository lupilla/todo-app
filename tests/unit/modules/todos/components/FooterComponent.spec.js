import { shallowMount } from "@vue/test-utils";
import FooterComponent from "@/modules/todos/components/FooterComponent.vue";
import {
  ACTIVE_STATE,
  COMPLETED_STATE,
  INITIAL_STATE,
} from "../../../mocks/state";
import createVuexStore from "../../../mocks/mock-store.js";
import router from "@/router";
const store = createVuexStore();
const $route = {
  path: "/",
  name: "allTodos",
};
describe("FooterComponent tests", () => {
  let wrapper;

  beforeEach(async () => {
    wrapper = shallowMount(FooterComponent, {
      global: {
        mocks: {
          $route,
        },
        plugins: [store, router],
      },
      props: {
        todos: INITIAL_STATE.todos,
      },
      computed: {
        getAllTodos: () => INITIAL_STATE.todos,
        getCompletedTodos: () =>
          INITIAL_STATE.todos.filter((todo) => todo.completed),
        getActiveTodos: () =>
          INITIAL_STATE.todos.filter((todo) => !todo.completed),
      },
    });
  });

  test("FooterComponent Component should match with snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("should show 'clear component' button with completed todos", () => {
    const clearButton = wrapper.find('[data-id-test="remove"]');
    expect(clearButton.exists()).toBe(true);
  });

  test("property 'linkSelected' should have 'all' value by default", () => {
    expect(wrapper.vm.$data.linkSelected).toBe("all");
  });

  test("should show 'clear component' button when all todos are completed", async () => {
    const $route = {
      path: "/completed",
      name: "completedTodos",
    };
    const wrapper = await shallowMount(FooterComponent, {
      global: {
        mocks: {
          $route,
        },
        plugins: [store, router],
      },
      props: {
        todos: COMPLETED_STATE.todos,
      },
      computed: {
        getAllTodos: () => COMPLETED_STATE.todos,
        getCompletedTodos: () =>
          COMPLETED_STATE.todos.filter((todo) => todo.completed),
        getActiveTodos: () =>
          COMPLETED_STATE.todos.filter((todo) => !todo.completed),
      },
    });
    const clearButton = wrapper.find('[data-id-test="remove"]');
    expect(clearButton.exists()).toBe(true);
    expect(clearButton.isVisible()).toBe(true);
  });

  test("should show FooterComponent properly when there are not todos", async () => {
    const wrapper = shallowMount(FooterComponent, {
      global: {
        mocks: {
          $route,
        },
        plugins: [store, router],
      },
      props: {
        todos: [],
      },
      computed: {
        getAllTodos: () => [],
        getCompletedTodos: () => [],
        getActiveTodos: () => [],
      },
    });
    expect(wrapper.isVisible()).toBe(true);
  });

  test("should show three buttons for filtering", async () => {
    const buttons = await wrapper.findAll("router-link-stub");
    expect(buttons.length).toBe(3);
  });
  test("should call 'selectedEvent' with 'all' argument when 'all' button is clicked", async () => {
    const spyNavigate = jest.spyOn(wrapper.vm, "selectedEvent");
    const buttons = await wrapper.findAll("router-link-stub");
    await buttons[0].trigger("click");
    expect(spyNavigate).toHaveBeenCalledWith("all");
  });

  test("should call 'selectedEvent' with 'active' argument when 'active' button is clicked", async () => {
    const spyNavigate = jest.spyOn(wrapper.vm, "selectedEvent");
    const buttons = await wrapper.findAll("router-link-stub");
    await buttons[1].trigger("click");
    expect(spyNavigate).toHaveBeenCalledWith("active");
  });

  test("should call 'selectedEvent' with 'completed' argument when 'completed' button is clicked", async () => {
    const spyNavigate = jest.spyOn(wrapper.vm, "selectedEvent");
    const buttons = await wrapper.findAll("router-link-stub");
    await buttons[2].trigger("click");
    expect(spyNavigate).toHaveBeenCalledWith("completed");
  });

  test("should change 'linkSelected' property with 'all' value when 'all' button is clicked", async () => {
    const buttons = await wrapper.findAll("router-link-stub");
    await buttons[0].trigger("click");
    expect(wrapper.vm.$data.linkSelected).toBe("all");
  });

  test("should change 'linkSelected' property with 'active' value when 'active' button is clicked", async () => {
    const buttons = await wrapper.findAll("router-link-stub");
    await buttons[1].trigger("click");
    expect(wrapper.vm.$data.linkSelected).toBe("active");
  });

  test("should change 'linkSelected' property with 'completed' value when 'completed' button is clicked", async () => {
    const buttons = await wrapper.findAll("router-link-stub");
    await buttons[2].trigger("click");
    expect(wrapper.vm.$data.linkSelected).toBe("completed");
  });

  test("'removeTodos' method should call 'deleteCompletedTodos' action", async () => {
    const spyRemoveTodos = jest.spyOn(wrapper.vm, "deleteCompletedTodos");
    wrapper.vm.removeTodos();
    expect(spyRemoveTodos).toHaveBeenCalled();
  });

  test("should call 'linkSelected' method when 'all' button is clicked", async () => {
    const spyselectedEvent = jest.spyOn(wrapper.vm, "selectedEvent");
    const buttons = await wrapper.findAll("router-link-stub");
    await buttons[0].trigger("click");
    expect(spyselectedEvent).toHaveBeenCalled();
  });

  test("should call 'linkSelected' method when 'active' button is clicked", async () => {
    const spyselectedEvent = jest.spyOn(wrapper.vm, "selectedEvent");
    const buttons = await wrapper.findAll("router-link-stub");
    await buttons[1].trigger("click");
    expect(spyselectedEvent).toHaveBeenCalled();
  });

  test("should call 'linkSelected' method when 'completed' button is clicked", async () => {
    const spyselectedEvent = jest.spyOn(wrapper.vm, "selectedEvent");
    const buttons = await wrapper.findAll("router-link-stub");
    await buttons[2].trigger("click");
    expect(spyselectedEvent).toHaveBeenCalled();
  });

  test("should call 'removeTodos' method when 'clear completed' button is clicked", async () => {
    const spyRemoveTodos = jest.spyOn(wrapper.vm, "removeTodos");
    const clearButton = wrapper.find('[data-id-test="remove"]');
    await clearButton.trigger("click");
    expect(spyRemoveTodos).toHaveBeenCalled();
  });

  test("property 'linkSelected' should have 'active' value when route is '/active'", () => {
    const $route = {
      path: "/active",
      name: "activeTodos",
    };
    const wrapper = shallowMount(FooterComponent, {
      global: {
        mocks: {
          $route,
        },
        plugins: [store, router],
      },
      props: {
        todos: ACTIVE_STATE.todos,
      },
      computed: {
        getAllTodos: () => ACTIVE_STATE.todos,
        getCompletedTodos: () =>
          ACTIVE_STATE.todos.filter((todo) => todo.completed),
        getActiveTodos: () =>
          ACTIVE_STATE.todos.filter((todo) => !todo.completed),
      },
    });
    expect(wrapper.vm.$data.linkSelected).toBe("active");
  });

  test("property 'linkSelected' should have 'completed' value when route is '/completed'", () => {
    const $route = {
      path: "/completed",
      name: "completedTodos",
    };
    const wrapper = shallowMount(FooterComponent, {
      global: {
        mocks: {
          $route,
        },
        plugins: [store, router],
      },
      props: {
        todos: COMPLETED_STATE.todos,
      },
      computed: {
        getAllTodos: () => COMPLETED_STATE.todos,
        getCompletedTodos: () =>
          COMPLETED_STATE.todos.filter((todo) => todo.completed),
        getActiveTodos: () =>
          COMPLETED_STATE.todos.filter((todo) => !todo.completed),
      },
    });
    expect(wrapper.vm.$data.linkSelected).toBe("completed");
  });
});
