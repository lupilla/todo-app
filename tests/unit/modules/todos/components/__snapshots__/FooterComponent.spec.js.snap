// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FooterComponent tests FooterComponent Component should match with snapshot 1`] = `
<footer class="footer">
  <div class="todo-count"><strong>[
  {
    "id": "-N0yGtCoVxtQyt39fNqF",
    "name": "go to the library",
    "completed": false,
    "isEditing": false
  },
  {
    "id": "-N0yWX2HtsbkDP1tuzf0",
    "name": "telephone the dentist",
    "completed": false,
    "isEditing": false
  }
] items left </strong></div>
  <ul class="filters">
    <li>
      <router-link-stub to="/" replace="false" custom="false" ariacurrentvalue="page" data-id-test="all" class="selected"></router-link-stub>
      <router-link-stub to="/active" replace="false" custom="false" ariacurrentvalue="page" data-id-test="active" class=""></router-link-stub>
      <router-link-stub to="/completed" replace="false" custom="false" ariacurrentvalue="page" data-id-test="completed" class=""></router-link-stub>
    </li>
  </ul><button type="button" data-id-test="remove" class="clear-completed"> Clear completed </button>
</footer>
`;
